### PR TITLE
Security queries are not implemented to spec

### LINKED ISSUE #2024: Security queries are not implemented to spec
### Bug Report
With the acceptance of [SDL 0317](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0317-sdl-protocol-security-specification.md), this library's implementation of security queries does not match the spec fully. Several issues exist with the current implementation such as:

1. When a security query is received from Core, its header is entirely ignored and the library assumes that it is a `Send Handshake Data` request. This causes issues if Core were to send another type of query, such as a `Send Internal Error` notification.

```objc
    // Tear off the binary header of the client protocol message to get at the actual TLS handshake
    // TODO: (Joel F.)[2016-02-15] Should check for errors
    NSData *clientHandshakeData = [clientHandshakeMessage.payload subdataWithRange:NSMakeRange(12, (clientHandshakeMessage.payload.length - 12))];
```

2. Hardcoded values are used when sending security queries for fields such as query type and query ID. Any predefined values should be properly added as constants.

```objc
    serverTLSPayload.functionID = 0x01; // TLS Handshake message
    serverTLSPayload.rpcType = 0x00;
    serverTLSPayload.correlationID = 0x00;
```

should be changed to something like

```objc
    serverTLSPayload.queryID = SDLQueryIDSendHandshakeData;
    serverTLSPayload.queryType = SDLQueryTypeResponse;
    serverTLSPayload.sequenceNumber = requestHeader.sequenceNumber;
```

3. Security queries are constructed using an RPC header, which has a slightly different format than a security query header. Security query headers should be implemented separate from RPC headers.

```objc
    // For a control service packet, we need a binary header with a function ID corresponding to what type of packet we're sending.
    SDLRPCPayload *serverTLSPayload = [[SDLRPCPayload alloc] init];
    ...
    NSData *binaryData = serverTLSPayload.data;

    return [SDLProtocolMessage messageWithHeader:serverMessageHeader andPayload:binaryData];
```

should be changed to something like

```objc
    // For a control service packet, we need a binary header with a function ID corresponding to what type of packet we're sending.
    SDLQueryPayload *serverTLSPayload = [[SDLQueryPayload alloc] init];
    ...
    NSData *binaryData = serverTLSPayload.data;

    return [SDLProtocolMessage messageWithHeader:serverMessageHeader andPayload:binaryData];
```

##### OS & Version Information
* iOS Version: N/A
* SDL iOS Version: master
* Testing Against: N/A

### PR DESCRIPTION
**Pull request by FrankElias77**
Fixes #2024 

### Risk
This PR makes **no** API changes.

### Testing Plan
- [ ] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Unit tests added for the new class added SDLSecurityQueryPayload

#### Core Tests
Triggered one of the errors in the proposal https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0317-sdl-protocol-security-specification.md#5123-error-codes. I sent an invalid queryId/queryType in the handshake response which prompted an error from core

Core version / branch / commit hash / module tested against: 7.1.1
HMI name / version / branch / commit hash / module tested against: generic_hmi/0.10.0

### Summary
Update how the security query header is being constructed.
Check security query header for error messages, for ex if core were to send an "internal error" notification

##### Breaking Changes
* N/A

##### Enhancements
* N/A

##### Bug Fixes
* Update how the security query header is being constructed instead of the previous SDLRPCPayload class.
Check the security query header to handle error cases, for ex if core were to send an "internal error" notification

### Tasks Remaining:
N/A

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+});
+
+describe(@"Data Tests", ^ {
+    it (@"Should convert to byte data correctly", ^ {
+        expect(testPayload.data).to(equal(testData()));
+    });
+});
+
+describe(@"RPCPayloadWithData Test", ^ {
+    it (@"Should convert from byte data correctly", ^ {
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:68]:
```suggestion
    it(@"should convert from byte data correctly", ^{
```

Fix all misformatting like this

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+QuickSpecBegin(SDLSecurityQueryPayloadSpec)
+
+__block SDLSecurityQueryPayload* testPayload;
+__block NSDictionary* dict = @{SDLRPCParameterNameResponse:
+                                   @{SDLRPCParameterNameParameters:@{},
+                                     SDLRPCParameterNameOperationName:SDLRPCFunctionNameDeleteCommand}};
+
+NSData* (^testData)(void) = ^NSData* {
+    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:dict options:0 error:0];
+    NSData* binaryData = [NSData dataWithBytes:"PrimitiveString" length:strlen("PrimitiveString")];
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:28]: We don't want to test an RPC like this. We want to test something like what we'd really see with JSON string and error code and binary data like we'll see.

**Review by joeljfischer — CHANGES_REQUESTED** [review]

**Review by joeljfischer — CHANGES_REQUESTED** [review]

**Review by joeljfischer — CHANGES_REQUESTED** [review]

**Review by joeljfischer — APPROVED** [review]

**Review by joeljfischer — CHANGES_REQUESTED** [review]

File: SmartDeviceLink/public/SDLSecurityQueryErrorCode.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
+//
```
Comment (joeljfischer) [inline SmartDeviceLink/public/SDLSecurityQueryErrorCode.h:1]: This new file was placed in the "public" folder, but it's a project file and non-public, so this project no longer works in Cocoapods. Please move this class' files to the "private" folder and fix the project.

**Review by joeljfischer — CHANGES_REQUESTED** [review]

**Review by joeljfischer — APPROVED** [review]

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    testPayload.queryType = 0x20;
+    testPayload.queryID = 0x02;
+    testPayload.sequenceNumber = 0x1443;
+    testPayload.jsonData = [NSJSONSerialization dataWithJSONObject:dict options:0 error:0];
+    testPayload.binaryData = [NSData dataWithBytes:"PrimitiveString" length:strlen("PrimitiveString")];
+});
+
+describe(@"Getter/Setter Tests", ^ {
+    it(@"should set and get correctly", ^ {
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:51]: Please remove whitespaces in blocks:  `^ {` ==> `^{`
There are a few more such cases I do not highlight each one.

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+#import <Foundation/Foundation.h>
+
+#import <Quick/Quick.h>
+#import <Nimble/Nimble.h>
+
+#import "SDLRPCParameterNames.h"
+#import "SDLRPCFunctionNames.h"
+#import "SDLSecurityQueryPayload.h"
+#import "SDLSecurityQueryErrorCode.h"
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:18]: Please reorder files alphabetically.

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    testPayload.binaryData = [NSData dataWithBytes:"PrimitiveString" length:strlen("PrimitiveString")];
+});
+
+describe(@"Getter/Setter Tests", ^ {
+    it(@"should set and get correctly", ^ {
+        expect(@(testPayload.queryType)).to(equal(SDLSecurityQueryTypeNotification));
+        expect(@(testPayload.queryID)).to(equal(SDLSecurityQueryIdSendInternalError));
+        expect(@(testPayload.sequenceNumber)).to(equal(@0x1443));
+        expect([NSJSONSerialization JSONObjectWithData:testPayload.jsonData options:0 error:0]).to(equal(dict));
+        expect([NSString stringWithUTF8String:[testPayload binaryData].bytes]).to(equal(@"PrimitiveString"));
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:56]: Be careful when converting bytes to strings and back again. Generally you should avoid using bytes directly. 
```
NSString *stringData = [[NSString alloc] initWithData:[testPayload binaryData] encoding:NSUTF8StringEncoding];
```
Also I propose to replace string literal to a const like following:
```
static NSString *const primitiveString = @"PrimitiveString";
```
You can use this string in all other places.

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+#import <Nimble/Nimble.h>
+
+#import "SDLRPCParameterNames.h"
+#import "SDLRPCFunctionNames.h"
+#import "SDLSecurityQueryPayload.h"
+#import "SDLSecurityQueryErrorCode.h"
+
+QuickSpecBegin(SDLSecurityQueryPayloadSpec)
+
+__block SDLSecurityQueryPayload* testPayload;
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:22]:
```
__block SDLSecurityQueryPayload *testPayload = nil;
```
Please nullify the pointer.

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+__block NSDictionary* dict = @{@"id": @"3", @"text": @"SDL does not support encryption"};
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:23]: Please do not use `__block` here, it is confusing. The `dict` is not going to change in this test suite. Also could you please dub it somehow differently, for instance `testDictionary`

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+NSData* (^testData)(void) = ^NSData* {
+    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:dict options:0 error:0];
+    NSData* binaryData = [NSData dataWithBytes:"PrimitiveString" length:strlen("PrimitiveString")];
+
+    Byte header[12] = {0x20, 0x00, 0x00, 0x02, 0x00, 0x00, 0x14, 0x43, 0x00, 0x00, 0x00, 0x00};
+    *(UInt32 *)&header[8] = CFSwapInt32HostToBig((unsigned int)jsonData.length);
+
+    NSMutableData *data = [NSMutableData dataWithCapacity:12 + jsonData.length];
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:32]: If you want to allocate data buffer beforehand then you had better do it a bit differently:
```
NSMutableData *data = [NSMutableData dataWithCapacity:sizeof(header) + jsonData.length + binaryData.length];
```

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+#import "SDLSecurityQueryErrorCode.h"
+
+QuickSpecBegin(SDLSecurityQueryPayloadSpec)
+
+__block SDLSecurityQueryPayload* testPayload;
+__block NSDictionary* dict = @{@"id": @"3", @"text": @"SDL does not support encryption"};
+
+NSData* (^testData)(void) = ^NSData* {
+    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:dict options:0 error:0];
+    NSData* binaryData = [NSData dataWithBytes:"PrimitiveString" length:strlen("PrimitiveString")];
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:27]: Be careful with string encoding especially avoid using raw bytes and repeating string literals, I would propose to make it other way round:
```
NSData *binaryData = [primitiveString dataUsingEncoding:NSUTF8StringEncoding];
```
And you can define `primitiveString` once and use it everywhere.

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+    Byte header[12] = {0x20, 0x00, 0x00, 0x02, 0x00, 0x00, 0x14, 0x43, 0x00, 0x00, 0x00, 0x00};
+    *(UInt32 *)&header[8] = CFSwapInt32HostToBig((unsigned int)jsonData.length);
+
+    NSMutableData *data = [NSMutableData dataWithCapacity:12 + jsonData.length];
+    [data appendBytes:&header length:12];
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:33]: Please avoid using magic numbers, we can do it a bit differently:
```
[data appendBytes:&header length:sizeof(header)];
```

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    it(@"should convert from byte data correctly", ^ {
+        SDLSecurityQueryPayload* constructedPayload = [SDLSecurityQueryPayload securityPayloadWithData:testData()];
+
+        expect(@(constructedPayload.queryType)).to(equal(SDLSecurityQueryTypeNotification));
+        expect(@(constructedPayload.queryID)).to(equal(SDLSecurityQueryIdSendInternalError));
+        expect(@(constructedPayload.sequenceNumber)).to(equal(@0x1443));
+        NSDictionary *jsonDict = [NSJSONSerialization JSONObjectWithData:constructedPayload.jsonData options:0 error:0];
+        expect(jsonDict).to(equal(dict));
+        expect(jsonDict[@"text"]).to(equal(SDLSecurityQueryErrorCodeNotSupported));
+        expect([NSString stringWithUTF8String:[constructedPayload binaryData].bytes]).to(equal(@"PrimitiveString"));
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:76]: I would propose to use other string function:
```
NSString *stringData = [[NSString alloc] initWithData:[testPayload binaryData] encoding:NSUTF8StringEncoding];
expect(stringData).to(equal(primitiveString));
```

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    });
+});
+
+QuickSpecEnd
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:80]: Please implement one more test for the following method:
```
- (instancetype)initWithQueryType:(SDLSecurityQueryType)queryType queryID:(SDLSecurityQueryId)queryID sequenceNumber:(UInt32)sequenceNumber jsonData:(nullable NSData *)jsonData binaryData:(nullable NSData *)binaryData;
```

File: SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+};
+
+beforeSuite(^{
+    testPayload = [[SDLSecurityQueryPayload alloc] init];
+
+    testPayload.queryType = 0x20;
+    testPayload.queryID = 0x02;
+    testPayload.sequenceNumber = 0x1443;
+    testPayload.jsonData = [NSJSONSerialization dataWithJSONObject:dict options:0 error:0];
+    testPayload.binaryData = [NSData dataWithBytes:"PrimitiveString" length:strlen("PrimitiveString")];
```
Comment (yoooriii) [inline SmartDeviceLinkTests/RPCSpecs/PayloadSpecs/SDLSecurityQueryPayloadSpec.m:47]: Please use string data `primitiveString` defined in one place without repeating itself:
```
testPayload.binaryData = [primitiveString dataUsingEncoding:NSUTF8StringEncoding];
```

**Review by yoooriii — COMMENTED** [review]
Would you please implement unit tests in SDLProtocolSpec since the protocol has changed and the updated code is not covered with tests. Also could you please review Codecov messages for it? Thank you.

**Comment by joeljfischer** [comment]
Hello @yoooriii, this is not the kind of review that we are looking for from the steering committee. Please place your review comments on the steering committee discussion (https://github.com/smartdevicelink/sdl_ios/discussions/2030) and limit your comments to whether the changes in the PR continue to support your head units' security capabilities. Thank you.

**Pull request merged**